core/data/src/main/java/uk/gov/gchq/gaffer/data/generator/OpenCypherCsvElementGenerator.java

```diff
@@ -0,0 +1,255 @@
+import uk.gov.gchq.koryphe.impl.function.ToDouble;
+import uk.gov.gchq.koryphe.impl.function.ToFloat;
+import uk.gov.gchq.koryphe.impl.function.ToInteger;
+import uk.gov.gchq.koryphe.impl.function.ToLong;
+import uk.gov.gchq.koryphe.impl.function.ToString;
+import uk.gov.gchq.koryphe.tuple.Tuple;
+
+import java.io.Serializable;
+
+import java.util.List;
+import java.util.Map;
+
+@Since("2.0.0")
+@Summary("Generates elements from an openCypher CSV string")
+public class OpenCypherCsvElementGenerator implements ElementGenerator<String>, Serializable {
+    private static final long serialVersionUID = -821376598172364516L;
+    private static final Logger LOGGER = LoggerFactory.getLogger(OpenCypherCsvElementGenerator.class);
+
+    public static final String VERTEX = ":ID";
+    public static final String NEO4J_VERTEX = "_id";
+    public static final String ENTITY_GROUP = ":LABEL";
+    public static final String NEO4J_ENTITY_GROUP = "_labels";
+
+    public static final String SOURCE = ":START_ID";
+    public static final String NEO4J_SOURCE = "_start";
+    public static final String DESTINATION = ":END_ID";
+    public static final String NEO4J_DESTINATION = "_end";
+
+    public static final String EDGE_GROUP = ":TYPE";
+    public static final String NEO4J_EDGE_GROUP = "_type";
+    private static final List<String> ELEMENT_COLUMN_NAMES = ImmutableList.of(VERTEX, ENTITY_GROUP, EDGE_GROUP, SOURCE, DESTINATION);
+    private String header;
+    private int firstRow = 0;
+    private Boolean trim = false;
+    private char delimiter = ',';
+    private String nullString = "";
+
+    @Override
+    public Iterable<? extends Element> apply(final Iterable<? extends String> strings) {
+        CsvLinesToMaps parseCsv = new CsvLinesToMaps()
+            .firstRow(1)
+            .trim(trim)
+            .nullString(nullString)
+            .delimiter(delimiter)
+            .parseHeader(header);
+
+        IterableFunction<Map<String, Object>, Tuple<String>> toTuples = new IterableFunction<>(new MapToTuple<String>());
+
+        FunctionChain.Builder<Tuple<String>, Tuple<String>> transformTuplesBuilder = new FunctionChain.Builder<>();
+
+        ElementTupleDefinition entityDefinition = new ElementTupleDefinition(ENTITY_GROUP).vertex(VERTEX);
+        ElementTupleDefinition edgeDefinition = new ElementTupleDefinition(EDGE_GROUP)
+            .source(SOURCE)
+            .destination(DESTINATION)
+            .property("edge-id", VERTEX);
+        for (final String columnHeader : parseCsv.getHeader()) {
+            if (!ELEMENT_COLUMN_NAMES.contains(columnHeader)) {
+                String propertyName = columnHeader.split(":")[0];
+                if (!PropertiesUtil.isValidName(propertyName)) {
+                    propertyName = propertyName.replaceAll("_", "-");
+                    propertyName = PropertiesUtil.stripInvalidCharacters(propertyName);
+                }
+                if (columnHeader.contains(":")) {
```

Comment:
> One of the Javadoc headers says "must be valid openCypher CSV format", but "openCypher CSV format" is not a well known standard at all, rather it is [a data format used exclusively by AWS Nepture's Bulk Loader](https://docs.aws.amazon.com/neptune/latest/userguide/bulk-load-tutorial-format-opencypher.html). It isn't used anywhere else, although it does resemble [the format used by Neo4J for export](https://neo4j.com/labs/apoc/4.1/export/csv/) (but with types). The Javadoc should link the the AWS page to make it clear what that format is.
> 
> I'm also concerned about the datatypes being part of the Column. It feels like it would be more useful to be able to import from a standard CSV [the way that Neo4j does](https://neo4j.com/docs/cypher-manual/current/clauses/load-csv/#load-csv-import-data-from-a-csv-file), with the AWS format as an optional extra. Probably best to add this as an improvement issue - with this added the class could be renamed to just 'CypherCSV'.
> 

You're right the format should be linked, but effectively the openCypher format is just a codified version of neo4j's csv format, with the added optional feature of types. As it currently works you could import files that have types, or do not (in Gaffer will default to String), and it also accepts both the openCypher standard of naming headers as well as the slightly different neo4j specific one. The main point is that it will work with both Neptune and neo4j, but this should be stated explicitly.


> It would also be good if there was an issue in gaffer-doc to cover documenting how to use this, as the Javadoc alone is not enough.


Issue tracked: https://github.com/gchq/gaffer-doc/issues/297